File: android/src/main/java/com/cap/plugins/capstatusbar/CAPStatusBarPlugin.java
package com.cap.plugins.capstatusbar;

import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.util.Objects;

@CapacitorPlugin(name = "CAPStatusBar")
public class CAPStatusBarPlugin extends Plugin {
    private static final String TAG = "CAPStatusBarPlugin";
    private final CAPStatusBar implementation = new CAPStatusBar();

    @Override
    public void load() {
        super.load();
        // Apply default style based on system theme on plugin load
        getActivity().runOnUiThread(() -> {
            implementation.applyDefaultStyle(getActivity());
        });
    }

    @PluginMethod
    public void setStyle(PluginCall call) {
        try {
            String style = call.getString("style");
            String color = call.getString("color");
            if (style == null) {
                call.reject("style is required");
                return;
            }
            getActivity().runOnUiThread(() -> {
                implementation.setStyle(getActivity(), style, color);
                call.resolve();
            });
        } catch (Exception e) {
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void show(PluginCall call) {
        try {
            boolean animated = Boolean.TRUE.equals(call.getBoolean("animated", true));
            getActivity().runOnUiThread(() -> {
                implementation.showStatusBar(getActivity(), animated);
                call.resolve();
            });
        } catch (Exception e) {
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void hide(PluginCall call) {
        try {
            boolean animated = Boolean.TRUE.equals(call.getBoolean("animated", true));
            getActivity().runOnUiThread(() -> {
                implementation.hideStatusBar(getActivity(), animated);
                call.resolve();
            });
        } catch (Exception e) {
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void setOverlaysWebView(PluginCall call) {
        try {
            Boolean value = call.getBoolean("value");
            if (value == null) {
                call.reject("value is required");
                return;
            }
            getActivity().runOnUiThread(() -> {
                implementation.setOverlaysWebView(getActivity(), value);
                call.resolve();
            });
        } catch (Exception e) {
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void setBackground(PluginCall call) {
        try {
            String color = call.getString("color");
            if (color == null) {
                call.reject("color is required");
                return;
            }
            getActivity().runOnUiThread(() -> {
                implementation.setBackground(getActivity(), color);
                call.resolve();
            });
        } catch (Exception e) {
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void getSafeAreaInsets(PluginCall call) {
        try {
            getActivity().runOnUiThread(() -> {
                java.util.Map<String, Integer> insets = implementation.getSafeAreaInsets(getActivity());
                com.getcapacitor.JSObject result = new com.getcapacitor.JSObject();
                result.put("top", insets.get("top"));
                result.put("bottom", insets.get("bottom"));
                result.put("left", insets.get("left"));
                result.put("right", insets.get("right"));
                call.resolve(result);
            });
        } catch (Exception e) {
            call.reject(e.getMessage());
        }
    }
}

File: ios/Sources/CAPStatusBarPlugin/CAPStatusBarPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(CAPStatusBarPlugin)
public class CAPStatusBarPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CAPStatusBarPlugin"
    public let jsName = "CAPStatusBar"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "setStyle", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "show", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "hide", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setOverlaysWebView", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setBackground", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getSafeAreaInsets", returnType: CAPPluginReturnPromise)
    ]
    private let implementation = CAPStatusBar()

    override public func load() {
        super.load()
        // Apply default style based on system theme on plugin load
        implementation.applyDefaultStyle()
    }

    @objc func setStyle(_ call: CAPPluginCall) {
        guard let style = call.getString("style") else {
            call.reject("style is required")
            return
        }
        let color = call.getString("color")
        implementation.setStyle(style: style, colorHex: color)
        call.resolve()
    }

    @objc func show(_ call: CAPPluginCall) {
        let animated = call.getBool("animated") ?? true
        implementation.show(animated: animated)
        call.resolve()
    }

    @objc func hide(_ call: CAPPluginCall) {
        let animated = call.getBool("animated") ?? true
        implementation.hide(animated: animated)
        call.resolve()
    }

    @objc func setOverlaysWebView(_ call: CAPPluginCall) {
        let value = call.getBool("value") ?? true
        implementation.setOverlaysWebView(value: value)
        call.resolve()
    }

    @objc func setBackground(_ call: CAPPluginCall) {
        guard let color = call.getString("color") else {
            call.reject("color is required")
            return
        }
        implementation.setBackground(colorHex: color)
        call.resolve()
    }

    @objc func getSafeAreaInsets(_ call: CAPPluginCall) {
        implementation.getSafeAreaInsets { insets in
            call.resolve([
                "top": insets["top"] ?? 0,
                "bottom": insets["bottom"] ?? 0,
                "left": insets["left"] ?? 0,
                "right": insets["right"] ?? 0
            ])
        }
    }
}
